pkg/autoscaler/autoscaler.go

```diff
@@ -283,116 +176,99 @@ func (a *Autoscaler) Scale(ctx context.Context, now time.Time) (int32, bool) {
+		logger.Info("Un-panicking.")
+		a.panicTime = nil
+		a.maxPanicPods = 0
 	}
 
 	var desiredPodCount int32
-
-	if a.panicking {
+	if a.panicTime != nil {
 		logger.Debug("Operating in panic mode.")
 		a.reporter.ReportPanic(1)
+		// We do not scale down while in panic mode. Only increases will be applied.
 		if desiredPanicPodCount > a.maxPanicPods {
-			logger.Infof("Increasing pods from %v to %v.", observedPanicPods, int(desiredPanicPodCount))
+			logger.Infof("Increasing pods from %v to %v.", readyPods, desiredPanicPodCount)
 			a.panicTime = &now
 			a.maxPanicPods = desiredPanicPodCount
 		}
-		desiredPodCount = int32(math.Ceil(a.maxPanicPods))
+		desiredPodCount = a.maxPanicPods
 	} else {
 		logger.Debug("Operating in stable mode.")
 		a.reporter.ReportPanic(0)
-		desiredPodCount = int32(math.Ceil(desiredStablePodCount))
+		desiredPodCount = desiredStablePodCount
 	}
 
 	a.reporter.ReportDesiredPodCount(int64(desiredPodCount))
 	return desiredPodCount, true
 }
 
-func (a *Autoscaler) aggregateData(now time.Time, stableWindow, panicWindow time.Duration) (*totalAggregation, *totalAggregation, map[string]Stat) {
+func (a *Autoscaler) aggregateData(now time.Time, stableWindow, panicWindow time.Duration) (float64, float64, statsBucket, bool) {
 	a.statsMutex.Lock()
 	defer a.statsMutex.Unlock()
 
-	// 60 second window
-	stableData := newTotalAggregation(stableWindow)
-
-	// 6 second window
-	panicData := newTotalAggregation(panicWindow)
+	var (
+		stableBuckets float64
+		stableTotal   float64
 
-	// Last stat per Pod
-	lastStat := make(map[string]Stat)
+		panicBuckets float64
+		panicTotal   float64
 
-	// Accumulate stats into their respective buckets
-	for key, stat := range a.stats {
-		instant := key.time
-		if instant.Add(panicWindow).After(now) {
-			panicData.aggregate(stat)
+		lastBucketTime time.Time
+		lastBucket     statsBucket
+	)
+	for bucketTime, bucket := range a.bucketed {
+		if !bucketTime.Add(panicWindow).Before(now) {
+			panicBuckets++
+			panicTotal += bucket.concurrency()
 		}
-		if instant.Add(stableWindow).After(now) {
-			stableData.aggregate(stat)
-
-			// If there's no last stat for this pod, set it
-			if _, ok := lastStat[stat.PodName]; !ok {
-				lastStat[stat.PodName] = stat
-			} else if lastStat[stat.PodName].Time.Before(*stat.Time) {
-				// If the current last stat is older than the new one, override
-				lastStat[stat.PodName] = stat
-			}
+
+		if !bucketTime.Add(stableWindow).Before(now) {
+			stableBuckets++
+			stableTotal += bucket.concurrency()
 		} else {
-			// Drop metrics after 60 seconds
-			delete(a.stats, key)
+			delete(a.bucketed, bucketTime)
+		}
+
+		if bucketTime.After(lastBucketTime) {
+			lastBucketTime = bucketTime
+			lastBucket = bucket
 		}
 	}
-	return stableData, panicData, lastStat
+
+	if stableBuckets == 0 {
+		return 0, 0, nil, false
+	}
+	return stableTotal / stableBuckets, panicTotal / panicBuckets, lastBucket, true
```

Comment:
`panicBuckets` could be zero. Do we need to check this?

Reply:
Shouldn't if stableBuckets is non-zero. I'll add a doublecheck nonetheless.

Reply:
Say there is only a pod which has been terminating 30 seconds ago. There are stats before 30 seconds ago, but no stats in the last 30 seconds. In this case `panicBuckets` is 0 while `stableBuckets` is not.